### ISSUE TITLE
Problem with numerical

### ISSUE THREAD

**Issue by pwiegers** [GBonnaire/jspreadsheet-plugins-and-editors#5]
Hi,

I'm trying to get jSpreadsheet to accept EU-style mumbers, like 1,25 instead of 1.25. (Having a thousand-separator using a . instead of a comma would be a bonus :-) )
I've been in contact with the official supprot of jSpeadsheet. While helpfull, they have not offered a complete solution. With their help I've been able to change the updateTable to display comma's instead of dots, but that does not work if you enter a formula:"= 1,2 * 3" will give you 6, instead of 2,6.
I was hoping that your numeric editor would solve this, but it suffers from the same issue.
I'm tryintg to solve this here: https://jsfiddle.net/pwiegers/yrso3jcL/1/
I think there should be some way to change text on save (ie: comma -> dot) [I got that far..] and then somehow trap the /result/ of the formula and change the dots back to comma's, but I don't know how to perform the last step.

Can you help me please?

regards,

Paul

**Comment by GBonnaire** [comment]
hi Paul,

Yes it's possible with custom editor

Try this custom editor (upgrade of my currency Editor)

```javascript
jspreadsheet.editors["text-eu"] = function() {
    var methods = {};
    var editor = null;
    var thousandSeparator = "";
    var decimalSeparator = ".";

    methods.createCell = function(cell, value, x, y, instance, options) {
        _init(options);
        cell.innerText = methods.parseValue(x, y, value, instance, options);
    }

    methods.updateCell = function(cell, value, x, y, instance, options) {
        if (cell) {
            cell.innerText = methods.parseValue(x, y, value, instance, options);
        }
    }

    methods.openEditor = function(cell, value, x, y, instance, options) {  
        editor = instance.parent.input;  
        editor.innerText = value;
        editor.focus();
    }

    methods.closeEditor = function(cell, save, x, y, instance, options) {
        if (save) {
            var value = editor.innerText;
        } else {
            var value = '';
        }

        return value;
    }

    methods.parseValue = function(x, y, value, instance, options) {
        // Formula?
        if ((''+value).substr(0,1) == '=' && instance.parent.config.parseFormulas == true) {
            value = instance.executeFormula(_parseFormula(value, options), x, y);
        }

        // Mask
        if(value!=="") {
            value = _getValueProcessed(value, options);
        }

        return value;
    }

    methods.get = function(options, value) {
        _init(options);
        return _getValueProcessed(value, options);
    }

    function _parseFormula(value, options) {
        var rg = new RegExp("([0-9\\"+thousandSeparator+"\\"+decimalSeparator+"]+)", "gm");
        return value.toString().replace(rg, _getValueNumber);
    }
    function _getValueProcessed(value, options) {
        if(value == "#ERROR") {
            return value;
        }
        if(typeof value == "string") {
            value = _getValueNumber(value);
        }
        return new Intl.NumberFormat(options.locale, options).format(value);
    }

    function _getValueNumber(value) {
        return parseFloat(value.toString()
            .replace(new RegExp('\\' + thousandSeparator, 'g'), '')
            .replace(new RegExp('\\' + decimalSeparator), '.')
        );
    }

    function _init(options) {
        thousandSeparator = Intl.NumberFormat(options.locale).format(11111).replace(/\p{Number}/gu, '');
        decimalSeparator = Intl.NumberFormat(options.locale).format(1.1).replace(/\p{Number}/gu, '');
    }

    return methods;
}();
```

to load in header avec jspreadsheet.js

example to use, it's important to defined locale property for use good format :

 ```javascript
   var MyTable = jspreadsheet(document.getElementById('spreadsheet'), {
        worksheets: [{
           minDimensions:[10,10],
           columns: [
               { type: "text-eu", locale: 'nl-NL' },
           ]
        }],
    });
```

Try that, and keep me touch if it's behavior that you search

**Comment by pwiegers** [comment]
Hi Guillaume,

thank you very much for your elaborate answer!
It _almost_ works...  if I enter =1,2*3 it works beatifully. But if I keep calculating with the result, I get NaN or #error as a result.
(also, if I enter text in this column, I get NaN, but I think I can correct that myself. Not so sure about the error above!)

I've implemented it like this: https://jsfiddle.net/pwiegers/cqekh3p2/16/

I trust that is correct?

regards,

Paul

**Comment by GBonnaire** [comment]
Try with this new version

```javascript
jspreadsheet.editors["text-eu"] = function() {
    var methods = {};
    var editor = null;
    var thousandSeparator = "";
    var decimalSeparator = ".";

    methods.createCell = function(cell, value, x, y, instance, options) {
        _init(options);
        cell.innerText = methods.parseValue(x, y, value, instance, options);
    }

    methods.updateCell = function(cell, value, x, y, instance, options) {
        if (cell) {
            cell.innerText = methods.parseValue(x, y, value, instance, options);
        }
    }

    methods.openEditor = function(cell, value, x, y, instance, options) {  
        editor = instance.parent.input;  
        editor.innerText = value;
        editor.focus();
    }

    methods.closeEditor = function(cell, save, x, y, instance, options) {
        if (save) {
            var value = editor.innerText;
        } else {
            var value = '';
        }

        return value;
    }

    methods.parseValue = function(x, y, value, instance, options) {
        // Formula?
        if ((''+value).substr(0,1) == '=' && instance.parent.config.parseFormulas == true) {
            value = instance.executeFormula(_parseFormula(value, options), x, y);
        }
        
        // Apply Mask
        if(value!=="" && _isNumber(value)) {
            value = _getValueProcessed(value, options);
        } else {
            value = jspreadsheet.editors.text.parseValue(x, y, value, instance);
        }
        
        return value;
    }

    methods.get = function(options, value) {
        _init(options);
        if(value!=="" && _isNumber(value)) {
            return _getValueProcessed(value, options);
        } else {
            return jspreadsheet.editors.text.get(options, value);
        }
    }

    function _isNumber(value) {
        var rg = new RegExp("([^0-9\\"+thousandSeparator+"\\"+decimalSeparator+"]+)", "gm");
        return value.toString().trim().replace(rg, "") == value.toString().trim();
    }
    
    function _parseFormula(value, options) {
        var rg = new RegExp("([0-9\\"+thousandSeparator+"\\"+decimalSeparator+"]+)", "gm");
        return value.toString().replace(rg, _getValueNumber);
    }
    function _getValueProcessed(value, options) {
        if(typeof value == "string") {
             value = _getValueNumber(value);
        }
        return new Intl.NumberFormat(options.locale, options).format(value);
    }

    function _getValueNumber(value) {
        return parseFloat(value.toString()
            .replace(new RegExp('\\' + thousandSeparator, 'g'), '')
            .replace(new RegExp('\\' + decimalSeparator), '.')
        );
    }

    function _init(options) {
        thousandSeparator = Intl.NumberFormat(options.locale).format(11111).replace(/\p{Number}/gu, '');
        decimalSeparator = Intl.NumberFormat(options.locale).format(1.1).replace(/\p{Number}/gu, '');
    }

    return methods;
}();
```

with that :
```
on A1 => 1,2*3 => 1,2*3
on A2 => =1,2*3 => 3,6
on B2 => =A2 => 3,6
```

**Comment by pwiegers** [comment]
You are amazing!
Thank you very much! :-)
This seems to work just like I need it too!

**Comment by GBonnaire** [comment]
My pleasure 😊

**Comment by GBonnaire** [comment]
I have create a new editor based on your problem here : https://github.com/GBonnaire/jspreadsheet-plugins-and-editors/blob/master/editors/text/jss.editor.textlocale.js

**Comment by pwiegers** [comment]
Hi Guillaume,

sorry to bother you again, but I find it very hard to follow all the calls, as some are hidden from me :-(
I appreciate the work you have done so far, but there is one problem remaining:
if in A1 I enter just "1,2" (meaning 1.2) and in A2 I enter: "=A1 * 3" I get NaN.
if in A1 I enter "=1,2" it will work correctly.
I cannot see where the difference is made in the code...

Can you help me one more time? :-)

regards,

Paul

**Comment by GBonnaire** [comment]
Hi,

Sorry, it's my fault, i miss a step in my previous code (Oups)

Try with new code editor :

```javascript
jspreadsheet.editors["text-eu"] = function() {
    var methods = {};
    var editor = null;
    var thousandSeparator = "";
    var decimalSeparator = ".";

    methods.createCell = function(cell, value, x, y, instance, options) {
        _init(options);
        return methods.updateCell(cell, value, x, y, instance, options);
    }

    methods.updateCell = function(cell, value, x, y, instance, options) {
        if (cell) {
            var v = methods.parseValue(x, y, value, instance, options);
            cell.innerText = v;
            if(_isNumber(v)) {
                return _getValueNumber(v);
            }
        }
    }

    methods.openEditor = function(cell, value, x, y, instance, options) {  
        editor = instance.parent.input;  
        editor.innerText = value;
        editor.focus();
    }

    methods.closeEditor = function(cell, save, x, y, instance, options) {
        if (save) {
            var value = editor.innerText;
        } else {
            var value = '';
        }

        return value;
    }

    methods.parseValue = function(x, y, value, instance, options) {
        // Formula?
        if ((''+value).substr(0,1) == '=' && instance.parent.config.parseFormulas == true) {
            value = instance.executeFormula(_parseFormula(value, options), x, y);
        }
        
        // Apply Mask
        if(value!=="" && _isNumber(value)) {
            value = _getValueProcessed(value, options);
        } else {
            value = jspreadsheet.editors.text.parseValue(x, y, value, instance);
        }
        
        return value;
    }

    methods.get = function(options, value) {
        _init(options);
        if(value!=="" && _isNumber(value)) {
            return _getValueProcessed(value, options);
        } else {
            return jspreadsheet.editors.text.get(options, value);
        }
    }

    function _isNumber(value) {
        var rg = new RegExp("([^0-9\\"+thousandSeparator+"\\"+decimalSeparator+"]+)", "gm");
        return value.toString().trim().replace(rg, "") == value.toString().trim();
    }
    
    function _parseFormula(value, options) {
        var rg = new RegExp("([0-9\\"+thousandSeparator+"\\"+decimalSeparator+"]+)", "gm");
        return value.toString().replace(rg, _getValueNumber);
    }
    function _getValueProcessed(value, options) {
        if(typeof value == "string") {
             value = _getValueNumber(value);
        }
        return new Intl.NumberFormat(options.locale, options).format(value);
    }

    function _getValueNumber(value) {
        return parseFloat(value.toString()
            .replace(new RegExp('\\' + thousandSeparator, 'g'), '')
            .replace(new RegExp('\\' + decimalSeparator), '.')
        );
    }

    function _init(options) {
        thousandSeparator = Intl.NumberFormat(options.locale).format(11111).replace(/\p{Number}/gu, '');
        decimalSeparator = Intl.NumberFormat(options.locale).format(1.1).replace(/\p{Number}/gu, '');
    }

    return methods;
}();
```

or this file : https://github.com/GBonnaire/jspreadsheet-plugins-and-editors/blob/master/editors/text/jss.editor.textlocale.js (name editor is different but code is same)

**Comment by pwiegers** [comment]
wow, you are fast!
<thank you!>

**Comment by GBonnaire** [comment]
My pleasure 😉 if it work, that is important 😉

**Comment by pwiegers** [comment]
Hi Guillaume,

I feel like I'm stalking you - I do not mean to! :-(
However, it is still not as it could be: it works, but if I change the source cell of a formula (ie: A1:1, A2:2, A3: =SUMA1:A2 and I change A2 to 3) it will not update the result of the formula...

Again thanks for your time. This appears to be a multi-itteration problem :-)

regards,

Paul

**Comment by GBonnaire** [comment]
Hi,

Hum ok, i make a fix and make all check (lol) now it's good 😊 

```javascript
jspreadsheet.editors["text-eu"] = function() {
    var methods = {};
    var editor = null;
    var thousandSeparator = "";
    var decimalSeparator = ".";

    methods.createCell = function(cell, value, x, y, instance, options) {
        _init(options);
        return methods.updateCell(cell, value, x, y, instance, options);
    }

    methods.updateCell = function(cell, value, x, y, instance, options) {
        if (cell) {
            cell.innerText = methods.parseValue(x, y, value, instance, options);
            if(value!="" && _isNumber(value)) {
                return _getValueNumber(value);
            }
        }
    }

    methods.openEditor = function(cell, value, x, y, instance, options) {  
        editor = instance.parent.input;  
        if(typeof value == "number") {
            value = _getValueProcessed(value, options);
        }
        editor.innerText = value;
        editor.focus();
    }

    methods.closeEditor = function(cell, save, x, y, instance, options) {
        if (save) {
            var value = editor.innerText;
        } else {
            var value = '';
        }

        return value;
    }

    methods.parseValue = function(x, y, value, instance, options) {
        // Formula?
        if ((''+value).substr(0,1) == '=' && instance.parent.config.parseFormulas == true) {
            value = instance.executeFormula(_parseFormula(value, options), x, y);
        }
        
        // Apply Mask
        if(value!=="" && _isNumber(value)) {
            value = _getValueProcessed(value, options);
        } else {
            value = jspreadsheet.editors.text.parseValue(x, y, value, instance);
        }
        
        return value;
    }

    methods.get = function(options, value) {
        _init(options);
        if(value!=="" && _isNumber(value)) {
            return _getValueProcessed(value, options);
        } else {
            return jspreadsheet.editors.text.get(options, value);
        }
    }

    function _isNumber(value) {
        var rg = new RegExp("([^\\-0-9\\"+thousandSeparator+"\\"+decimalSeparator+"]+)", "gm");
        return value.toString().trim().replace(rg, "") == value.toString().trim();
    }
    
    function _parseFormula(value, options) {
        var rg = new RegExp("([0-9\\"+thousandSeparator+"\\"+decimalSeparator+"]+)", "gm");
        return value.toString().replace(rg, _getValueNumber);
    }
    function _getValueProcessed(value, options) {
        if(typeof value == "string") {
             value = _getValueNumber(value);
        }
        return new Intl.NumberFormat(options.locale, options).format(value);
    }

    function _getValueNumber(value) {
        return parseFloat(value.toString()
            .replace(new RegExp('\\' + thousandSeparator, 'g'), '')
            .replace(new RegExp('\\' + decimalSeparator), '.')
        );
    }

    function _init(options) {
        thousandSeparator = Intl.NumberFormat(options.locale).format(11111).replace(/\p{Number}/gu, '');
        decimalSeparator = Intl.NumberFormat(options.locale).format(1.1).replace(/\p{Number}/gu, '');
    }

    return methods;
}();
```

**Comment by pwiegers** [comment]
Hi Guillaume,

I've implemented your code in my app. I've made a small change:
```
    function _isNumber(value) {
        var rg = new RegExp("([^-0-9\\"+thousandSeparator+"\\"+decimalSeparator+"]+)", "gm");
        return value.toString().trim().replace(rg, "") == value.toString().trim();
    }
```
in order for negative numbers to be recognized as, well, numbers.
I also have a follow-up question: if I enter a formula, and _drag_ the little square down, the formula gets copied to the next cell - but not updated. (IE: =A1+B1 should update to =A2+B2, but stays the same. )
This has to have something to do with this code, as the spreadsheet normally does this correctly, but I do not know where... Could you come to the rescue once more? :-)

regards,

Paul

**Comment by GBonnaire** [comment]
Hi,

Thanks for your help, i update my code on previous message.

For the drag formula, i look/search why

**Comment by GBonnaire** [comment]
Hi,

An option in column can add this feature : `shiftFormulas`

on column properties set shiftFormulas to true and check

available only on JSS Pro v8.1.8 and more

**Comment by pwiegers** [comment]
Hi,

I downloaded the latest version from jspreadsheet.com (not clear what version?)
but the shiftFormula does not seem to change anything...

regards,

Paul

**Comment by GBonnaire** [comment]
Hi,

The last version isn't already online. That will be on new version

For check, you go on www.jspreadsheet.com open your console and execute this command :

`jspreadsheet.version()`

Regards

**Comment by pwiegers** [comment]
Hi,

I am on 8.1.7, indeed...
Is there a URL where I can download 8.1.8?

regards,

Paul

**Comment by GBonnaire** [comment]
Hi,

the next version with this feature is not online. Editor of JSS will publish soon. 

Please wait 😊
